### PR TITLE
Refactor ClassFactory to avoid trying certain APIs on known android versions that don't support them

### LINKED ISSUE #949: 1.9.0 - UnsupportedOperationException when deserialising java class with non-empty constructor
We have (had) some legacy java DTO code:
```java
public class ProblemDTO {
    private String message;

    ProblemDTO(String message) {
        this.message = message;
    }
}
```

The Moshi crashes on API 29 when trying to deserialise it:
```
 Caused by: java.lang.UnsupportedOperationException: ObjectStreamClass.getConstructorId(Class<?>) is not supported on SDK 28
        at java.io.ObjectStreamClass.getConstructorId(ObjectStreamClass.java:2421)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.squareup.moshi.ClassFactory.get(ClassFactory.java:110)
        at com.squareup.moshi.ClassJsonAdapter$1.create(ClassJsonAdapter.java:96)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:138)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:98)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:72)
```

An example project is here - https://github.com/emartynov/MoshiTest

The fix for the issue was the conversion of java DTO to Kotlin.

### PR DESCRIPTION
**Pull request by ZacSweers**
This improves the android support in `ClassFactory` to not try grey-listed reflection APIs on 28+, and also gates gingerbread-only APIs to only try below API 10. This leverages some of the prior android support we'd added to support it.

Resolves #949, sort of. The workaround posted in it is the right solution, this just makes it not try something we know won't work.

### CODE REVIEW

**Comment by ZacSweers** [comment]
I'll try this in an app targeting API 28+ to test this.

File: moshi/src/main/java/com/squareup/moshi/ClassFactory.java [hunk 1]
```diff
@@ -98,57 +90,91 @@ public T newInstance() throws InvocationTargetException, IllegalAccessException
...
-        @SuppressWarnings("unchecked")
-        @Override public T newInstance() throws InvocationTargetException, IllegalAccessException {
-          return (T) newInstance.invoke(null, rawType, Object.class);
-        }
-        @Override public String toString() {
-          return rawType.getName();
-        }
-      };
-    } catch (Exception ignored) {
+    if (isBelowAndroidApi(10)) {
```
Comment (ZacSweers) [inline moshi/src/main/java/com/squareup/moshi/ClassFactory.java]:
```suggestion
    if (isBelowAndroidApi(9)) {
```

File: moshi/src/main/java/com/squareup/moshi/ClassFactory.java [hunk 2]
```diff
@@ -98,57 +90,91 @@ public T newInstance() throws InvocationTargetException, IllegalAccessException
...
-        @SuppressWarnings("unchecked")
-        @Override public T newInstance() throws InvocationTargetException, IllegalAccessException {
-          return (T) newInstance.invoke(null, rawType, Object.class);
-        }
-        @Override public String toString() {
-          return rawType.getName();
-        }
-      };
-    } catch (Exception ignored) {
+    if (isBelowAndroidApi(9)) {
```
Comment (ZacSweers) [inline moshi/src/main/java/com/squareup/moshi/ClassFactory.java]: Note I can't test this because it doesn't seem like it's possible to download pre-JB android versions anymore via ADB manager

File: moshi/src/main/java/com/squareup/moshi/ClassFactory.java [hunk 3]
```diff
@@ -57,18 +59,8 @@ abstract T newInstance() throws
...
-        int sdk = (int) sdkIntField.get(null);
-        if (sdk >= 29) {
-          tryUnsafe = false;
-        }
-      } catch (Exception ignored) {
-      } finally {
-        androidChecked = true;
-      }
+    if (!androidChecked && isAndroid()) {
+      tryUnsafe = isBelowAndroidApi(29);
```
Comment (dlew) [inline moshi/src/main/java/com/squareup/moshi/ClassFactory.java]: The code is setting up a complex volatile state here that is a bit hard to follow. E.g., one might wonder (as I did) why it's not just calling `isAndroid()`, since that method  *also* checks `!androidChecked` ahead of time.

What if we just sussed all these variables out at once? Like, write a method that resolves `androidChecked`, `isAndroid`, `tryUnsafe`, and `androidSdkInt` in one fell swoop. Then the code becomes easier to follow.

File: moshi/src/main/java/com/squareup/moshi/ClassFactory.java [hunk 4]
```diff
@@ -98,57 +90,91 @@ public T newInstance() throws InvocationTargetException, IllegalAccessException
...
+          }
+          @Override public String toString() {
+            return rawType.getName();
+          }
+        };
+      } catch (Exception ignored) {
+      }
     }
 
     throw new IllegalArgumentException("cannot construct instances of " + rawType.getName());
```
Comment (dlew) [inline moshi/src/main/java/com/squareup/moshi/ClassFactory.java]: Since we have no solution for API 29+, maybe we should throw a more instructive exception in that case (pointing them to the workaround, i.e. "you are on API 29+, you need an empty constructor")?

File: moshi/src/main/java/com/squareup/moshi/ClassFactory.java [hunk 5]
```diff
@@ -98,57 +90,91 @@ public T newInstance() throws InvocationTargetException, IllegalAccessException
...
+        androidSdkInt = (int) sdkIntField.get(null);
+      } catch (Exception ignored) {
+      } finally {
+        androidChecked = true;
+      }
+    }
+    return isAndroid;
+  }
+
+  private static boolean isBelowAndroidApi(int target) {
```
Comment (dlew) [inline moshi/src/main/java/com/squareup/moshi/ClassFactory.java]: Minor, but I found this method name confusing - what is "is" here?

Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/ClassFactory.java]: Cleaned up in 415ffb0

File: moshi/src/main/java/com/squareup/moshi/ClassFactory.java (hunk 4 above)
Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/ClassFactory.java]: Good call - 40b2ed3

File: moshi/src/main/java/com/squareup/moshi/ClassFactory.java (hunk 3 above)
Comment (ZacSweers) [reply moshi/src/main/java/com/squareup/moshi/ClassFactory.java]: Good idea - e793ecf and dcc427c

**Comment by ZacSweers** [comment]
Going to do one more check on API 28 and API 27 devices before merging

**Comment by ZacSweers** [comment]
Abandoning for now since we opted to revert covering these

**Comment by dlew** [comment]
Just for future knowledge - the revert was in #964.
